pinot-core/src/main/java/org/apache/pinot/core/operator/query/LinearSelectionOrderByOperator.java

```diff
@@ -0,0 +1,476 @@
+     * starts or should start.
+     */
+    private int _sortedPrefixSize;
+    private final int _maxSize;
+    private PriorityQueue<Object[]> _queue = null;
+    private int _maxPriorityQueueValues;
+
+    public PartiallySortedListBuilder(int maxSize, Comparator<Object[]> partitionComparator,
+        Comparator<Object[]> unsortedComparator) {
+      Preconditions.checkArgument(maxSize > 0, "Max size must be greater than 0");
+      _maxSize = maxSize;
+      _rows = new ArrayList<>(maxSize);
+      _partitionComparator = partitionComparator;
+      _unsortedComparator = unsortedComparator;
+    }
+
+    @Override
+    public boolean add(Object[] row) {
+      boolean result;
+      if (_lastPartitionRow == null) {
+        _sortedPrefixSize = 0;
+        _lastPartitionRow = row;
+        _rows.add(row);
+        return false;
+      } else {
+        int cmp = _partitionComparator.compare(row, _lastPartitionRow);
+        if (cmp < 0) {
+          throw new IllegalArgumentException("Row " + Arrays.toString(row) + " is lower than previously added row"
+              + Arrays.toString(_lastPartitionRow));
+        }
+
+        if (_queue != null) {
+          if (cmp > 0) {
+            // the new row is in a greater partition, just don't add it and continue
+            return true;
+          }
+          // otherwise they are equal by the already sorted columns. Let's check the other columns and evict the higher
+          // also, we have the higher value at the last position
+          SelectionOperatorUtils.addToPriorityQueue(row, _queue, _maxPriorityQueueValues);
+          return false;
+        } else {
+          if (_rows.size() == _maxSize) { // we already have _maxSize results
+            if (cmp > 0) {
+              // and the new one is higher, so we don't need to add it
+              sortLastPartition();
+              return true;
+            }
+            // and the new one may be smaller, so this is going to be the last partition and we need to prepare the
+            // priority queue
+            buildPriorityQueue();
+            SelectionOperatorUtils.addToPriorityQueue(row, _queue, _maxPriorityQueueValues);
+            return false;
+          }
+          if (cmp > 0) {
+            sortLastPartition();
+            _lastPartitionRow = row;
+            result = true;
+          } else {
+            result = false;
+          }
+          _rows.add(row);
+          return result;
+        }
+      }
+    }
+
+    private void buildPriorityQueue() {
+      _maxPriorityQueueValues = _maxSize - _sortedPrefixSize;
+      _queue = new PriorityQueue<>(_maxPriorityQueueValues, _unsortedComparator.reversed());
+      for (int i = 0; i < _maxPriorityQueueValues; i++) {
+        int lastIdx = _rows.size() - 1;
+        _queue.add(_rows.get(lastIdx));
+        _rows.remove(lastIdx);
+      }
+    }
+
+    void sortLastPartition() {
```

Comment:
We do need to sort to fulfill the contract of the interface we implement. It may be true that in the specific case where we are using this class right now it may not be necessary but we cannot break the contract of an interface we use. Given that this is a new interface, we could its semantics, but I think that is not a good idea. It would make the semantics very unclear.

The cost of sorting the last part is negligible when compared with the improvements of this PR and the problems that may be associated with a very complex interface semantic

Reply:
Which interface are you referring to? There is no public interface, and everything should be performance driven. I agree the improvement of this PR is huge, but this sort can take majority of the time (check the difference between partial sort and full sort, which should be mostly contributed by it). The contract for the `ListBuilder` is also clear, essentially the sorted columns are guaranteed to be sorted, and the unsorted columns (partial sorted case) is not sorted

Reply:
So your idea is that `ListBuilder.build()` invariant should be that elements are ordered by partition, but not by the not already sorted columns. In the current implementation, I think that makes sense. I'm going to change it.

But I think we should think at longer term. My original idea was more ambitious than that. The idea was to sort each block in the map phase (therefore in parallel) so the reduce/combine phase (that is called by a single thread) should have a linear cost. Also, by applying the sorting in the map phase we can use the local knowledge derived from sorted columns and make the comparison faster by ignoring the fields that are already sorted in the partition.